### ISSUE TITLE
Script to stop outbound traffic

### ISSUE THREAD

**Issue by rdarioc** [ukanth/afwall#387]
Guys, I need your help.

I was doing some maintenance on my home router when I found out the following notification:

* * *
**UDP Flood to Host** 192.168.2.50, 36040->> 93.50.94.24, 61522
* * *
Where .50 was my android device (4.4.2) and the destination is somewhere in Milan, Italy. Now since these are the days of Galileo's spyware and HT news are all over, I need to:
1. find which app started that flood and
2. block all traffic toward that sub-net
possibly i could be able to do so with the scripting feature of AFWall+, but I don't know how.

I understand this may not be the best place to ask for such advice and I apologize for it, any suggestion or recommendation on where to better post this request for help will be greatly appreciate

**Comment by CHEF-KOCH** [comment]
We are not here to discuss about scripts, this is an issue tracker not an general usage talking place. You can use the XDA thread for this.

**Comment by rdarioc** [comment]
@CHEF-KOCH, thank you for your advice. Do you perhaps have a pointer to the appropriate thread you are talking about?

**Comment by an0n981** [comment]
http://forum.xda-developers.com/showthread.php?t=1957231
[4.0+][ROOT][2.0.0-BETA] AFWall+ IPTables Firewall [19 JULY 2015]

**Comment by rdarioc** [comment]
Thank you so much Stewie!!!

**Comment by CHEF-KOCH** [comment]
Or just [look over here](https://github.com/ukanth/afwall/wiki), most question are answered already.

**Comment by rdarioc** [comment]
I have to say before posting I already looked were you said but I didn't find any script to block outbound traffic,  only inbound.  Anyway I'll look now on XDA.  Of course if you already knew the answer it would be great if you wanted to share it

**Comment by CHEF-KOCH** [comment]
You don't need any script to block outgoing stuff, that's what AFWall+ was designed for. If any _spyware_ app or whatever you want to call it want's to connect it need the internet permission in the manifest, so and because of this you could use the whitelist mode (default) to only allow the checked apps allowing the outgoing communication - it means you can uncheck everything and onjly select the stuff you really need, step by step. This also allows to identify the apps that consume much or _suspect_ traffic, just take a look in the log section.

If you are more professional you can use NetworkLog/Inceptor-NG/Whireshark/Burp apps/software to identify which is the app that cause that and what exactly is sent during the session. 

Blocking a specific IP doesn't prevent anything btw, since most if not all malware using random c networks to not easily just block the domain/IP to stop that. So you really need to identify the app first and better uninstall it, another way could be to use apps like Titanium Backup and freeze/uninstall/stop xyz suspect apps one by one and see if the connection is still present. 


Please also note that some malware apps also only connecting via Wiki/VPN only to intelligent _hide_ from your eyes because if they consume more traffic over cellular networks the user would get more pay attention to what's going on - So I always suggest to test stuff under a Wifi environment.

**Comment by rdarioc** [comment]
Sorry if I took so long to reply, I wanted first to read all the blog posts suggested by Stewie (an0n9801), plus I had to send the family on vacation!

First, I solved the mentioned issue by adding the following script:
$IPTABLES -A "afwall" -d 93.50.94.24/32 -j DROP || exit

and replicating such rule for every host/net I wanted to cut off.

With regard to your more detailed notes, I admit those were not that clear to me for the following reasons:
* * *
> You don't need any script to block outgoing stuff, that's what AFWall+ was designed for. If any
> spyware app or whatever you want to call it want's to connect it need the internet permission in the
> manifest, so and because of this you could use the whitelist mode (default) to only allow the
> checked apps allowing the outgoing communication - it means you can uncheck everything and only
> select the stuff you really need, step by step. This also allows to identify the apps that consume
> much or suspect traffic, just take a look in the log section.
* * *
That would be true for those "spyware" that run as separate processes, but if one "apparently" normal app, or even a kernel module has been affected I cannot follow your procedure. Plus I wanted to block the access to specific addresses/subnets, whatever process was generating it.
In other words, more than a simple "user friendly" UI to a firewall, I really wanted something at a lower level.

* * *
> If you are more professional you can use NetworkLog/Inceptor-NG/Whireshark/Burp apps/software
> to identify which is the app that cause that and what exactly is sent during the session.
* * *
I am familiar with Whireshark but how can identify a process that is generating some packets toward a specific address by simply looking at those packets on a network sniffer?

I could identify the source, which I already knew being my phone, but a process running in it just by reading at the packets? That is mysterious to me.

As for the other notes of course I agree with your comments.

**Comment by CHEF-KOCH** [comment]
Kernel modules can't be infected ([that easy](http://blog.onyxbits.de/how-worried-should-you-be-about-the-nsas-plans-to-hijack-google-play-project-irritant-horn-89/) since the spyware app would need root permission for such scenario (MITM,...) + most of malware I know also needs to run as admin (which needs to be enabled by the user first) to prevent from been killed by other _security_ apps and to redirect the entire DNS traffic -> and for locking down Android (Ransomware [admin part here]). As I said it's more important to remove the app that is possible infected or suspect instead of just blocking the IP's because if a c network (cdn = content delivery network) is in usage the whole IP blocking thing doesn't have any effect (except you really know all the IP's (and most cdn's can't be blocked because that would break most of all pages) but that's almost impossible without deep package inception and capturing each request).

> $IPTABLES -A "afwall" -d 93.50.94.24/32 -j DROP || exit

Better use $IPT -A "afwall" -d 93.50.94.24/32 -j REJECT which also needs no || exit because that would give an error. The REJECT should be used instead of DROP for several reasons (ie ICMP destination unreachable "problems").

> I really wanted something at a lower level.

Iptables is already the lowest level you can get on Linux. Blocking cell towers or capturing such stuff needs apps or tools I can't mention here (and I never give tips on such things).

>  I am familiar with Whireshark but how can identify a process that is generating some packets toward a specific address by simply looking at those packets on a network sniffer? I could identify the source, which I already knew being my phone, but a process running in it just by reading at the packets? That is mysterious to me.

Well, Wireshark have a good FAQ page, see also [here](https://ask.wireshark.org/questions/26512/capture-android-wlan-traffic) and [here](https://ask.wireshark.org/questions/3087/capturing-mobile-phone-packets) - if you know the UID you just could use this to identify the packets that are coming from the app or process. 
An procedure would be to use interceptor-ng/networklog app to see what could pass the firewall or just create an isolated lan environment for testing what really wants to go out. 


> [cut].. but a process running in it just by reading at the packets? That is mysterious to me.

Not only for you, depending on which apk we're talking about here it may execute something before the real manifest is read (a lot of apps doing this). But I can't give any more details here without more information but we already hit the off-topic now.
